File: python_receiver/src/rtu_receiver/xtea.py
from __future__ import annotations

import struct

_DELTA = 0x9E3779B9
_MASK32 = 0xFFFFFFFF


def _key_to_words_le(key: bytes) -> tuple[int, int, int, int]:
    if len(key) != 16:
        raise ValueError("XTEA key must be 16 bytes")
    return struct.unpack("<4I", key)


def xtea_decrypt_block_le(block8: bytes, key16: bytes, rounds: int = 32) -> bytes:
    if len(block8) != 8:
        raise ValueError("XTEA block must be 8 bytes")
    if rounds <= 0:
        raise ValueError("rounds must be positive")

    v0, v1 = struct.unpack("<2I", block8)
    k0, k1, k2, k3 = _key_to_words_le(key16)
    key_words = (k0, k1, k2, k3)

    summation = (_DELTA * rounds) & _MASK32
    for _ in range(rounds):
        v1 = (v1 - ((((v0 << 4) ^ (v0 >> 5)) + v0) ^ (summation + key_words[(summation >> 11) & 3]))) & _MASK32
        summation = (summation - _DELTA) & _MASK32
        v0 = (v0 - ((((v1 << 4) ^ (v1 >> 5)) + v1) ^ (summation + key_words[summation & 3]))) & _MASK32

    return struct.pack("<2I", v0, v1)


def xtea_encrypt_block_le(block8: bytes, key16: bytes, rounds: int = 32) -> bytes:
    if len(block8) != 8:
        raise ValueError("XTEA block must be 8 bytes")
    if rounds <= 0:
        raise ValueError("rounds must be positive")

    v0, v1 = struct.unpack("<2I", block8)
    k0, k1, k2, k3 = _key_to_words_le(key16)
    key_words = (k0, k1, k2, k3)

    summation = 0
    for _ in range(rounds):
        v0 = (v0 + ((((v1 << 4) ^ (v1 >> 5)) + v1) ^ (summation + key_words[summation & 3]))) & _MASK32
        summation = (summation + _DELTA) & _MASK32
        v1 = (v1 + ((((v0 << 4) ^ (v0 >> 5)) + v0) ^ (summation + key_words[(summation >> 11) & 3]))) & _MASK32

    return struct.pack("<2I", v0, v1)


def xtea_decrypt_ecb_le(ciphertext: bytes, key16: bytes, rounds: int = 32) -> bytes:
    if not ciphertext or len(ciphertext) % 8 != 0:
        raise ValueError("ciphertext length must be positive and divisible by 8")

    out = bytearray()
    for i in range(0, len(ciphertext), 8):
        out.extend(xtea_decrypt_block_le(ciphertext[i : i + 8], key16, rounds=rounds))
    return bytes(out)


def xtea_encrypt_ecb_le(plaintext: bytes, key16: bytes, rounds: int = 32) -> bytes:
    if not plaintext or len(plaintext) % 8 != 0:
        raise ValueError("plaintext length must be positive and divisible by 8")

    out = bytearray()
    for i in range(0, len(plaintext), 8):
        out.extend(xtea_encrypt_block_le(plaintext[i : i + 8], key16, rounds=rounds))
    return bytes(out)

File: node_receiver/src/xtea.js
const DELTA = 0x9e3779b9 >>> 0;

function keyToWordsLE(key16) {
  if (key16.length !== 16) {
    throw new Error("XTEA key must be 16 bytes");
  }
  return [
    key16.readUInt32LE(0),
    key16.readUInt32LE(4),
    key16.readUInt32LE(8),
    key16.readUInt32LE(12),
  ];
}

export function xteaDecryptBlockLE(block8, key16, rounds = 32) {
  if (block8.length !== 8) {
    throw new Error("XTEA block must be 8 bytes");
  }
  if (rounds <= 0) {
    throw new Error("rounds must be positive");
  }

  let v0 = block8.readUInt32LE(0) >>> 0;
  let v1 = block8.readUInt32LE(4) >>> 0;
  const keyWords = keyToWordsLE(key16);

  let sum = Math.imul(DELTA, rounds) >>> 0;
  for (let i = 0; i < rounds; i += 1) {
    v1 = (v1 - ((((v0 << 4) ^ (v0 >>> 5)) + v0) ^ ((sum + keyWords[(sum >>> 11) & 3]) >>> 0))) >>> 0;
    sum = (sum - DELTA) >>> 0;
    v0 = (v0 - ((((v1 << 4) ^ (v1 >>> 5)) + v1) ^ ((sum + keyWords[sum & 3]) >>> 0))) >>> 0;
  }

  const out = Buffer.alloc(8);
  out.writeUInt32LE(v0 >>> 0, 0);
  out.writeUInt32LE(v1 >>> 0, 4);
  return out;
}

export function xteaEncryptBlockLE(block8, key16, rounds = 32) {
  if (block8.length !== 8) {
    throw new Error("XTEA block must be 8 bytes");
  }
  if (rounds <= 0) {
    throw new Error("rounds must be positive");
  }

  let v0 = block8.readUInt32LE(0) >>> 0;
  let v1 = block8.readUInt32LE(4) >>> 0;
  const keyWords = keyToWordsLE(key16);

  let sum = 0;
  for (let i = 0; i < rounds; i += 1) {
    v0 = (v0 + ((((v1 << 4) ^ (v1 >>> 5)) + v1) ^ ((sum + keyWords[sum & 3]) >>> 0))) >>> 0;
    sum = (sum + DELTA) >>> 0;
    v1 = (v1 + ((((v0 << 4) ^ (v0 >>> 5)) + v0) ^ ((sum + keyWords[(sum >>> 11) & 3]) >>> 0))) >>> 0;
  }

  const out = Buffer.alloc(8);
  out.writeUInt32LE(v0 >>> 0, 0);
  out.writeUInt32LE(v1 >>> 0, 4);
  return out;
}

export function xteaDecryptEcbLE(ciphertext, key16, rounds = 32) {
  if (ciphertext.length === 0 || ciphertext.length % 8 !== 0) {
    throw new Error("ciphertext length must be positive and divisible by 8");
  }

  const chunks = [];
  for (let i = 0; i < ciphertext.length; i += 8) {
    chunks.push(xteaDecryptBlockLE(ciphertext.subarray(i, i + 8), key16, rounds));
  }
  return Buffer.concat(chunks);
}

export function xteaEncryptEcbLE(plaintext, key16, rounds = 32) {
  if (plaintext.length === 0 || plaintext.length % 8 !== 0) {
    throw new Error("plaintext length must be positive and divisible by 8");
  }

  const chunks = [];
  for (let i = 0; i < plaintext.length; i += 8) {
    chunks.push(xteaEncryptBlockLE(plaintext.subarray(i, i + 8), key16, rounds));
  }
  return Buffer.concat(chunks);
}
